### PR TITLE
[flang][cuda] Get device address in fir.declare

### PR DESCRIPTION
**Pull request by clementval**
Add pattern that update fir.declare memref when it comes from a device global and is not a descriptor. In that case, we recover the device address that needs to be used in ops like `fir.array_coor` and so on.

### CODE REVIEW

**Review by wangzpgi — APPROVED** [review]

File: flang/test/Fir/CUDA/cuda-global-addr.mlir [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  %c1_i32 = arith.constant 1 : i32
+  %c0_i32 = arith.constant 0 : i32
+  %c10 = arith.constant 10 : index
+  %1 = fir.shape %c10 : (index) -> !fir.shape<1>
+  %3 = fir.address_of(@_QMmod1Eadev) : !fir.ref<!fir.array<10xi32>>
+  %4 = fir.declare %3(%1) {data_attr = #cuf.cuda<device>, uniq_name = "_QMmod1Eadev"} : (!fir.ref<!fir.array<10xi32>>, !fir.shape<1>) -> !fir.ref<!fir.array<10xi32>>
+  %5 = fir.alloca i32 {bindc_name = "i", uniq_name = "_QFEi"}
+  %6 = fir.declare %5 {uniq_name = "_QFEi"} : (!fir.ref<i32>) -> !fir.ref<i32>
+  fir.store %c0_i32 to %6 : !fir.ref<i32>
+  %7 = fir.array_coor %4(%1) %c4 : (!fir.ref<!fir.array<10xi32>>, !fir.shape<1>, index) -> !fir.ref<i32>
```
Comment (Renaud-K) [inline flang/test/Fir/CUDA/cuda-global-addr.mlir:21]: Is the issue that previously, we were not passing the device address to array_coor ? If so then maybe the test below can check for it.

Comment (clementval) [reply flang/test/Fir/CUDA/cuda-global-addr.mlir:21]: Yes the issue was that we were using the host address here. I can update the test.

Comment (Renaud-K) [reply flang/test/Fir/CUDA/cuda-global-addr.mlir:21]: It is a good change then. When I did it, I probably was just happy to have a basic flow. This should fix a lot of tests. Thank you!

Comment (clementval) [reply flang/test/Fir/CUDA/cuda-global-addr.mlir:21]: Yeah it only shows up when you have a subscript in the data transfer for example smth like: 

```
adev(I+4) = 1
```
When we use the whole array we didn't see this issue.

**Review by Renaud-K — APPROVED** [review]

**Pull request merged**